### DISCUSSION TITLE
Styled components - babel plugin - lesbare class names

### DISCUSSION THREAD

**Discussion by SLS-BLN** [neuefische/questions#16] [Web] [answered]
Eine kurze Frage zum React Template und Babel:

Ich würde gerne das **Babel Plugin für styled components** einsetzen. Es wird in den Docs von styled components sehr empfohlen:

"It's recommended (but not required) to also use the [styled-components Babel plugin](https://github.com/styled-components/babel-plugin-styled-components) if you can. It offers many benefits like more legible class names, server-side rendering compatibility, smaller bundles, and more."

Zur Installation muss auch die **babel configuration** geändert werden.

Hat jemand eine Idee, wie das geht genau geht? Und ob das überhaupt möglich ist? Babel ist in der package.json Datei selbst nicht installiert.

**Comment by actyralf** [comment, marked as answer]
Bei Apps, die mit create-react-app erstellt werden, ist Babel schon automatisch dabei (versteckt in den react-scripts). Die .babelrc kann in diesem Setup aber NICHT geändert werden. Wenn man das unbedingt machen muss, kann man das Projekt "ejecten" - das bedeutet dann aber, dass man sich um die komplette React-Konfiguration (Updates, Abhängigkeiten, Versionen etc.) selbst kümmern muss, das ist auch nicht reversibel.
Wenn ich das richtig sehe, hilft das Plugin, bestimmte Probleme beim Serverside Rendering zu beheben, das kommt bei uns/euch aber ja sowieso nicht zur Anwendung, oder? Wenn es keine konkreten Probleme gibt, würde ich daher an der Standard-Konfiguration nichts ändern. Hier noch Details zum Ejecten: [Don't eject your create-react-app](https://medium.com/curated-by-versett/dont-eject-your-create-react-app-b123c5247741)

**Reply by actyralf** [reply]
Update: Hiermit ist es wohl möglich, das plugin ohne eject zu verwenden: https://github.com/kentcdodds/babel-plugin-macros

**Reply by SLS-BLN** [reply]
Vielen Dank für die Info!

**Reply by actyralf** [reply]
> Vielen Dank für die Info!

Wenn die Antwort so ausreicht, bitte als "beantwortet" markieren - Danke!

**Reply by SLS-BLN** [reply]
## Eligible class names
Das Plugin ist installiert - noch ein Hinweis, warum es so wichtig ist:
Statt völlig unlesbarer class names wie ...

`<section class="sc-bczRLJ cgeka-d"><div>App goes here</div>;)</section>`

gibt das Plugin in den Browser Dev Tools dieses heraus:

`<section class="App__Wrapper-sc-3vbase-0 jwjPPq"><div>App goes here</div>;)</section>`

Die dazu gehörige App.js sieht so aus:

```import styled from 'styled-components/macro';

export default function App() {
  return (
    <Wrapper>
      <div>App goes here</div>;)
    </Wrapper>
  );
}

const Wrapper = styled.section`
  padding: 4em;
  background: papayawhip;
`; 
```

## Die Installation über npm erfolgt in zwei Schritten:

`npm install --save-dev babel-plugin-styled-components`
`npm install --save-dev babel-plugin-macros`

[Babel Macro](https://styled-components.com/docs/tooling#babel-macro)

**Reply by actyralf** [reply]
Eine Ergänzung: Alle Importe aus Styled Components (z.B. `styled`, `createGlobalStyle` etc.) dürfen dann **NICHT** mehr direkt aus `styled-components` importiert werden, sondern aus `styled-components/macro`, also z.B.
```
import styled from "styled-components/macro";
```
